### PR TITLE
api: rename `ht_enabled` to `smt`, make it optional and forbid setting it to True on ARM

### PR DESCRIPTION
**Pull request by alindima**
# Reason for This PR

We plan on renaming/removing ht_enabled in a future release and this is a first step towards this without introducing a breaking change in this release.
Also, making it required did not make sense for ARM, where it actually does nothing if configured. Added an error for when it's configured as True on ARM.

## Description of Changes

View commits

- [ ] This functionality can be added in [`rust-vmm`][1].

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes follow the [Runbook for Firecracker API changes][2].
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

**Comment by alindima** [comment]
There are some expected integration test failures caused by outdated artifacts in S3.
I will regenerate and upload them once I get some approvals on this PR

File: src/api_server/src/request/machine_configuration.rs
Comment (raduiliescu) [inline src/api_server/src/request/machine_configuration.rs:183]: Can we also test that both true and false fail on aarch64?

File: src/api_server/src/request/machine_configuration.rs
Comment (raduiliescu) [inline src/api_server/src/request/machine_configuration.rs]: Not clear why this were changed from true to false?

Comment (alindima) [reply src/api_server/src/request/machine_configuration.rs]: So that those unit tests don't fail on ARM

File: src/api_server/src/request/machine_configuration.rs
Comment (alindima) [reply src/api_server/src/request/machine_configuration.rs:183]: Actually only `ht_enabled=True` fails on ARM. `ht_enabled` defaults to False on all platforms.

**Comment by acatangiu** [comment]
Making `"ht_enabled": false` fail on ARM _is a breaking change_. Make sure to go through the versioned api support runbook :smile_cat:

**Comment by alindima** [comment]
> Making `"ht_enabled": false` fail on ARM _is a breaking change_. Make sure to go through the versioned api support runbook 😸

This PR only makes `ht_enabled: true` fail on ARM and makes this field optional in the machine-config.
I don't believe we should go down the deprecation path since this was actually a bug. (this field does nothing on ARM if set to True).
Moreover, we are including this in the v1.0 release, which allows for breaking changes since it's a major version bump.

File: src/api_server/src/request/machine_configuration.rs
Comment (georgepisaltu) [inline src/api_server/src/request/machine_configuration.rs:76]: I would disallow using this field on ARM altogether. What purpose does it serve to allow users to set it to false? It's just confusing.

```suggestion
    if vm_config.ht_enabled.is_some() {
```

If we go down this path, we could even have the `ht_enabled` field in the `VmConfig` struct marked with a `#[cfg(target_arch = "aarch64")]` so that we never mistakenly deserialize/serialize it.

Comment (alindima) [reply src/api_server/src/request/machine_configuration.rs:76]: I generally agree, but I was trying to minimise the breaking change impact. Let's discuss this some more

Comment (acatangiu) [reply src/api_server/src/request/machine_configuration.rs:76]: A breaking change is a breaking change is a breaking change.

Why do half the job to have only a breaking change in some cases? (setting `ht true on arm` will break, but `ht false on arm` will still work albeit it being useless/confusing)

Comment (alindima) [reply src/api_server/src/request/machine_configuration.rs:76]: We've discussed this topic inside the team and reached the following path forward:

keep the behaviour of this PR but additionally rename the `ht_enabled` param to `smt` (it will be optional and return error if set to True on ARM). This enables us to add support for 'emulated' smt in ARM as well if we need to in the future. We think this is the most extensible and cleanest option.

WDYT @acatangiu ?

Comment (alindima) [reply src/api_server/src/request/machine_configuration.rs]: I've implemented this in the latest commit

Comment (acatangiu) [reply src/api_server/src/request/machine_configuration.rs]: Sounds reasonable, but I don't see why you're not disabling/removing it altogether on ARM. You could seamlessly add it later with default value `false` when/if you ever implement smt emulation for arm. And in the meantime you're not left with a stub optional param which allows only false or otherwise fails the request.

You're basically keeping a button saying "this button might be useful later, but don't push it now or it will error" :smile:

Comment (alindima) [reply src/api_server/src/request/machine_configuration.rs]: > L.E.: an argument for keeping the stub is if that way the api requests would be identical between x86_64 and aarch64 -> more uniform code on customer side

This is one of the reasons I thought about.
It's also that otherwise, it may also be confusing for users to see that there is no ht_enabled config on arm. One may then wonder if it's emulated or not. The answer would be clearer if there was an ht_enabled option that defaults to `false` and that only allows False on Arm.

File: tests/integration_tests/functional/test_api.py
Comment (georgepisaltu) [inline tests/integration_tests/functional/test_api.py:184]: Why not `platform.machine() == 'x86_64'`?

File: tests/integration_tests/functional/test_api.py
Comment (georgepisaltu) [inline tests/integration_tests/functional/test_api.py:396]: Maybe we should split this test into multiple parts so that we get rid of this error. The test is indeed pretty long now.

**Review by georgepisaltu — APPROVED** [review]
PR looks great! Just a couple of minor comments.

**Review by luminitavoicu — APPROVED** [review]
Great job!

File: tests/integration_tests/functional/test_api.py
Comment (alindima) [reply tests/integration_tests/functional/test_api.py]: done

File: tests/integration_tests/functional/test_api.py
Comment (alindima) [reply tests/integration_tests/functional/test_api.py:396]: It's fairly long but I can't think of a way to purposefully split this test into multiple ones. It's a pretty cohesive one.

**Comment by alindima** [comment]
I've regenerated S3 artifacts and changed microvm config files to no longer include `ht_enabled` flag.
I've also performed a couple of changes in the test framework to work with the older firecracker artifacts.
The PR is currently using a temporary folder in the S3 bucket but once I see all tests are passing I'll move the artifacts and switch back the folder.

**Review by georgepisaltu — APPROVED** [review]

**Review by AlexandruCihodaru — APPROVED** [review]

**Review by luminitavoicu — APPROVED** [review]

**Pull request merged**
